.travis.yml

```diff
@@ -1,19 +1,18 @@
 sudo: false
 language: go
-go:
-  - 1.6.x
-  - 1.7.x
-  - 1.8.x
-  - 1.9.x
-  - master
 matrix:
   allow_failures:
     - go: master
   fast_finish: true
+  include:
+    - go: 1.10.x
+    - go: 1.11.x
+      env: GOFMT=1
+    - go: master
 install:
   - # Do nothing. This is needed to prevent default install action "go get -t -v ./..." from happening here (we want it to happen inside script step).
 script:
   - go get -t -v ./...
-  - diff -u <(echo -n) <(gofmt -d .)
+  - if test -n "${GOFMT}"; then gofmt -w -s . && git diff --exit-code; fi
```

Comment:
> What is the motivation for changing `diff -u <(echo -n) <(gofmt -d -s .)` to `gofmt -w -s . && git diff --exit-code`?

The old approach created a doubled diff, and the round of leading `+` it put before everything was distracting.  Compare [this][1] and [this][2].

[1]: https://travis-ci.org/gregjones/httpcache/jobs/442602930#L500
[2]: https://travis-ci.org/gregjones/httpcache/jobs/452142960#L506

Reply:
Thanks for explaining, that makes sense.

I'd like to look a way to continue to rely on `gofmt -d` to display a gofmt diff, rather than `gofmt -w` followed by `git diff`, if possible. The motivation is that:

1. It doesn't modify the source code. Even though `gofmt -w` should not change behavior, it's still better to avoid modifying source code in the general case (and trying to revert changes can make the command verbose).
2. It's not fallible to missing gofmt issues due to .gitignore or other git configuration details.

An idea is to use `-n` flag to `diff` rather than `-u`, since it produces a cleaner diff without the leading `+`s. I prototyped it in commit eedefab2b6fc226d2ccb892988c4e3e936b9a044. You can see its Travis output at https://travis-ci.org/gregjones/httpcache/jobs/452715773.

How do you feel about going with that approach?

Reply:
I'm not worried about either of your cases ;).  If someone submits an ignored file, I expext we'd reject that regardless of its Go formatting.  And if `gofmt` makes *any* change to an unignored file, the tests will fail regardless of whether the change is functional.  Who cares what's on disk when the failed container gets torn down?

But your output looks fine too, so I'm fine with you landing your approach instead.

Reply:
> Who cares what's on disk when the failed container gets torn down?

I don't mind what's on disk after the container is being shutdown. It's just that `gofmt -w` writes to disk first, and then the rest of the tests run on potentially modified .go files second.

I know that for this repo it's extremely unlikely it'd ever make a difference. However, I'd like to find a set of commands as general and simple as possible, in order to keep the Travis scripts used in many repositories in sync.

> But your output looks fine too, so I'm fine with you landing your approach instead.

Great! If you don't mind, please update this PR to use it. I think the rest of the changes look good so far.

Thanks!

Reply:
> It's just that `gofmt -w` writes to disk first, and then the rest of the tests run on potentially modified `.go` files second.

But the tests as a whole are going to fail anyway (because `git diff` will notice the changes).  So again, I don't see an issue.  But searching for a happy medium, how about I move the `gofmt` call to the end of the `script` block?  Would that address your concerns?

> However, I'd like to find a set of commands as general and simple as possible, in order to keep the Travis scripts used in many repositories in sync.

That's a good goal, but I don't see a lot of complexity or generality space between `gofmt -w -s . && git diff --exit-code` and `diff -n <(echo -n) <(gofmt -d .)`.

> > But your output looks fine too, so I'm fine with you landing your approach instead.
>
> Great! If you don't mind, please update this PR to use it.

I'm fine with a parallel PR landing instead of this one.  But as long as we're working on this one, I'd rather keep hunting until we find an approach that we both like.  I'm ok with the nested-diff approach (whether it uses `-u` or `-n`), but I don't *like* it ;).

Reply:
> But as long as we're working on this one, I'd rather keep hunting until we find an approach that we both like. I'm ok with the nested-diff approach (whether it uses `-u` or `-n`), but I don't _like_ it ;).

Understandable. If at any point you decide you'd rather just land this sooner, let me know. Until then, here are some more thoughts on this.

> But the tests as a whole are going to fail anyway (because git diff will notice the changes).

I was referring to the situation where `gofmt -w` writes first, `git diff` reports no diff (likely because nothing changed), and then tests run and succeed.

> But searching for a happy medium, how about I move the `gofmt` call to the end of the script block?

I would not prefer that more. I think running a `gofmt` check first is a good idea, since it's fast and a more basic check than all the tests. It's just that I want to find a way of running it without writing to disk.

`gofmt` itself isn't so bad; but I've seen other repos where the similar approach of "writing to repo and then using git diff" was used to check that `go generate` doesn't produce a diff. `go generate` may produce a diff, and I wouldn't want that to influence the rest of the tests. But trying to revert a potential diff means more `git` commands... Which is what motivates me to look for a solution that finds out whether there's a diff without writing to disk. That may not be possible in the case of `go generate` though.

Reply:
> > > It's just that `gofmt -w` writes to disk first, and then the rest of the tests run on potentially modified `.go` files second.
> >
> > But the tests as a whole are going to fail anyway (because `git diff` will notice the changes).
>
> I was referring to the situation where `gofmt -w` writes first, `git diff` reports no diff (likely because nothing changed), and then tests run and succeed.

Wait, that's what we want to happen for a successful PR, right?  If `gofmt` is happy with the originals, that's great, right?  I thought you were concerned about cases where the originals would have failed the unit tests (or some other later test), but the `gofmt -w` call adjusts the originals to get them to pass the later tests instead *and* `git diff` fails to notice the change vs. the originals.  I tried to rebut that situation [here][1], based on the unlikeliness of a gitignored `.go` file.  Are you still worried about gitignored `.go` files?  If so, they probably merit an explict CI check, because we should be rejecting them regardless of whether they are properly formatted Go ;).

> The idea of writing the .go files before testing them kinda bugs me.

But if we write something different, the test run as a whole will fail.  In the very unlikely case that the `gofmt` output passes a unit test that the originals would have failed, the gofmt failure will still keep the broken originals from landing.

> `go generate` may produce a diff, and I wouldn't want that to influence the rest of the tests.

Again, why does a pure test matter?  You need everything green before the code lands, so cross-talk in a run that's going to fail anyway doesn't seem like a big deal to me.

[1]: https://github.com/gregjones/httpcache/pull/92#discussion_r232277863

Reply:
Thanks for the discussion. I don't have further comments for now. I'm happy for this to go in, and I can keep looking for any potential improvements in the future.

Sorry for the delay, this PR slipped off my radar and no one pinged.

Reply:
> Sorry for the delay, this PR slipped off my radar and no one pinged.

No worries, I wasn't in a hurry ;).  Thanks for the merge :)